cosmos/client.go

```diff
@@ -173,7 +167,48 @@ func (c *Client) Stream(ctx context.Context, query string) (chan hash.Hash, chan
 	return hashC, errC, nil
 }
 
-// GetAccount returns the keybase's account.
-func (c *Client) GetAccount() (keys.Info, error) {
-	return c.kb.Get(c.accName)
+// GetAccount returns the local account.
+func (c *Client) GetAccount() (auth.Account, error) {
+	c.getAccountMutex.Lock()
+	defer c.getAccountMutex.Unlock()
+	if c.acc == nil {
+		accKb, err := c.kb.Get(c.accName)
+		if err != nil {
+			return nil, err
+		}
+		c.acc = auth.NewBaseAccount(
+			accKb.GetAddress(),
+			nil,
+			accKb.GetPubKey(),
+			AccNumber,
+			0,
+		)
+	}
+	localSeq := c.acc.GetSequence()
+	if accR, err := auth.NewAccountRetriever(c).GetAccount(c.acc.GetAddress()); err == nil {
+		c.acc = accR
+		// replace seq if sup
+		if localSeq > c.acc.GetSequence() {
+			c.acc.SetSequence(localSeq)
+		}
+	}
+	return c.acc, nil
+}
+
+// Sign signs a msg and return a tendermint tx.
+func (c *Client) sign(msg sdktypes.Msg) (tenderminttypes.Tx, error) {
+	acc, err := c.GetAccount()
+	if err != nil {
+		return nil, err
+	}
+
+	sequence := acc.GetSequence()
+	acc.SetSequence(acc.GetSequence() + 1)
```

Comment:
I'm not sure, but I think lock might be needed here. We retrive account and the between GetAccount and GetSequence someone might call it. Possible solution is to return sequence from a helper function GetAccountSeq or protect this with mutex.

Reply:
The lock is in the function `BuildAndBroadcastMsg` around the call of `sign`. This may not be the best but we plan to move all the account + sign logic to a dedicated structure.

Reply:
issue created: <https://github.com/mesg-foundation/engine/issues/1565>